### PR TITLE
[HW11] Add `UpdateTask`, `CreateUser` mutations

### PR DESCRIPTION
**Pull request by w1nway**
### Summary

A brief description of the pull request.

### How it works

Screenshots/screencasts of the pull request introduced functionality.

### Test plan

List of steps to manually test introduced functionality:

- Go to http://localhost:3000/graphiql
- Make request using schema:
```graphql
  query {
    me {
      id
      name
    }
  }
```

### Deploy notes

Notes regarding deployment the contained body of work.
These should note any db migrations, services, scripts, etc.

### Checklist:

- [ ] I have performed a self-review of my own code
- [ ] My pull request is shorter than 500 lines

### CODE REVIEW

File: app/graphql/mutations/create_user.rb [hunk 1]
```diff
@@ -0,0 +1,19 @@
+module Mutations
+  class CreateUser < BaseMutation
+    include GraphqlErrors
+
+    argument :input, Types::Inputs::CreateUserInput, required: true
+
+    type Types::Payloads::CreateUserPayload
+
+    def resolve(input:)
+      result = Users::Create.call(user_params: input.to_h, user: current_user)
```
Comment (RasimKhusaenov) [inline app/graphql/mutations/create_user.rb:10]: current_user?

File: app/graphql/mutations/update_task.rb [hunk 2]
```diff
@@ -0,0 +1,20 @@
...
+  class UpdateTask < BaseMutation
+    include GraphqlErrors
+
+    argument :input, Types::Inputs::UpdateTaskInput, required: true
+
+    type Types::Payloads::UpdateTaskPayload
+
+    def resolve(input:)
+      task = Task.find_by(id: input.task_id)
+      result = Tasks::Update.call(task: task, task_params: input.to_h.except(:task_id), user: current_user)
```
Comment (RasimKhusaenov) [inline app/graphql/mutations/update_task.rb:11]: current_user?

File: app/graphql/types/inputs/create_user_input.rb [hunk 3]
```diff
@@ -0,0 +1,9 @@
+module Types
+  module Inputs
+    class CreateUserInput < Types::BaseInputObject
+      argument :email, String, required: true
+      argument :first_name, String, required: false
+      argument :last_name, String, required: false
```
Comment (RasimKhusaenov) [inline app/graphql/types/inputs/create_user_input.rb:6]: password?

File: app/graphql/types/user_type.rb [hunk 4]
```diff
@@ -0,0 +1,14 @@
+# frozen_string_literal: true
+
+module Types
+  class UserType < Types::BaseObject
+    field :id, ID, null: false
+    field :first_name, String
+    field :last_name, String
+    field :email, String
+    field :password_digest, String
```
Comment (RasimKhusaenov) [inline app/graphql/types/user_type.rb:9]: Why we need it in type?

File: app/graphql/types/user_type.rb [hunk 5]
```diff
@@ -0,0 +1,14 @@
...
+    field :created_at, GraphQL::Types::ISO8601DateTime, null: false
+    field :updated_at, GraphQL::Types::ISO8601DateTime, null: false
+    field :role, String, null: false
```
Comment (RasimKhusaenov) [inline app/graphql/types/user_type.rb:12]: It is okay, but ideally it should be special type:
```ruby
module Types
  class UserRoleEnum < BaseEnum
    User.role.each_value do |user_role|
      value user_role.to_s.upcase, user_role.text, value: user_role.to_s
    end
  end
end
```

File: app/interactors/users/create.rb [hunk 6]
```diff
@@ -0,0 +1,20 @@
+module Users
+  class Create
+    include Interactor
+
+    delegate :user_params, to: :context
+
+    def call
+      context.user = user
+
+      # context.fail!(error: "Invalid data") unless user.update(user_params)
```
Comment (RasimKhusaenov) [inline app/interactors/users/create.rb:10]: Don't commit commented code

File: app/graphql/mutations/create_user.rb (hunk 1 above)
Comment (w1nway) [reply app/graphql/mutations/create_user.rb:10]: i was going to remove this

File: app/graphql/mutations/update_task.rb (hunk 2 above)
Comment (w1nway) [reply app/graphql/mutations/update_task.rb:11]: it too ))

File: app/graphql/types/user_type.rb (hunk 4 above)
Comment (w1nway) [reply app/graphql/types/user_type.rb:9]: it was generated with another fields

File: app/graphql/mutations/create_user.rb [hunk 7]
```diff
@@ -0,0 +1,19 @@
...
+
+    type Types::Payloads::CreateUserPayload
+
+    def resolve(input:)
+      result = Users::Create.call(user_params: input.to_h)
+
+      if result.success?
+        result.to_h
+      else
+        result.to_h.merge(errors: formatted_errors(result.user))
```
Comment (RasimKhusaenov) [inline app/graphql/mutations/create_user.rb:15]:
```suggestion
      result.to_h.merge(errors: formatted_errors(result.user))
```

File: app/graphql/mutations/create_user.rb [hunk 8]
```diff
@@ -0,0 +1,19 @@
+module Mutations
+  class CreateUser < BaseMutation
+    include AuthenticableGraphqlUser
```
Comment (RasimKhusaenov) [inline app/graphql/mutations/create_user.rb:3]: Why we need current user to create user?

File: app/graphql/types/payloads/create_user_payload.rb [hunk 9]
```diff
@@ -0,0 +1,8 @@
+module Types
+  module Payloads
+    class CreateUserPayload < Types::BaseObject
+      field :user, UserType, null: true
```
Comment (RasimKhusaenov) [inline app/graphql/types/payloads/create_user_payload.rb:4]: Can it be null?

**Review by RasimKhusaenov — CHANGES_REQUESTED** [review]

File: app/graphql/types/payloads/create_user_payload.rb (hunk 9 above)
Comment (w1nway) [reply app/graphql/types/payloads/create_user_payload.rb:4]: no, it my bad attentiveness again

File: app/graphql/mutations/update_task.rb [hunk 10]
```diff
@@ -0,0 +1,20 @@
...
+
+    def resolve(input:)
+      task = Task.find_by(id: input.task_id)
+      result = Tasks::Update.call(task: task, task_params: input.to_h.except(:task_id))
+
+      if result.success?
+        result.to_h
+      else
+        result.to_h.merge(errors: formatted_errors(result.task))
+      end
```
Comment (RasimKhusaenov) [inline app/graphql/mutations/update_task.rb:17]:
```suggestion
      result.to_h.merge(errors: formatted_errors(result.task))
```

File: app/graphql/types/payloads/update_task_payload.rb [hunk 11]
```diff
@@ -0,0 +1,8 @@
+module Types
+  module Payloads
+    class UpdateTaskPayload < Types::BaseObject
+      field :task, TaskType, null: true
```
Comment (RasimKhusaenov) [inline app/graphql/types/payloads/update_task_payload.rb:4]: Can it be null?

File: app/graphql/types/user_type.rb (hunk 5 above)
Comment (RasimKhusaenov) [reply app/graphql/types/user_type.rb:9]: @w1nway

**Review by RasimKhusaenov — CHANGES_REQUESTED** [review]

File: app/graphql/types/user_type.rb (hunk 5 above)
Comment (w1nway) [reply app/graphql/types/user_type.rb:9]: i think that question was closed

File: app/graphql/types/payloads/update_task_payload.rb [hunk 12]
```diff
@@ -0,0 +1,8 @@
+module Types
+  module Payloads
+    class UpdateTaskPayload < Types::BaseObject
+      field :task, TaskType, null: false
+      field :errors, [Types::UserError], null: true
```
Comment (RasimKhusaenov) [inline app/graphql/types/payloads/update_task_payload.rb:5]: Why `null: true`?

**Review by RasimKhusaenov — CHANGES_REQUESTED** [review]

File: app/graphql/types/payloads/create_project_payload.rb [hunk 13]
```diff
@@ -2,7 +2,7 @@ module Types
   module Payloads
     class CreateProjectPayload < Types::BaseObject
       field :project, ProjectType, null: false
-      field :errors, [Types::UserError], null: false
+      field :errors, [Types::UserError], null: true
```
Comment (RasimKhusaenov) [inline app/graphql/types/payloads/create_project_payload.rb:5]: Why?

File: app/graphql/types/payloads/create_user_payload.rb [hunk 14]
```diff
@@ -0,0 +1,8 @@
+module Types
+  module Payloads
+    class CreateUserPayload < Types::BaseObject
+      field :user, UserType, null: false
+      field :errors, [Types::UserError], null: true
```
Comment (RasimKhusaenov) [inline app/graphql/types/payloads/create_user_payload.rb:5]: Why `null: true`?

File: app/graphql/types/payloads/update_project_payload.rb [hunk 15]
```diff
@@ -2,7 +2,7 @@ module Types
   module Payloads
     class UpdateProjectPayload < Types::BaseObject
       field :project, Types::ProjectType, null: false
-      field :errors, [Types::UserError], null: false
+      field :errors, [Types::UserError], null: true
```
Comment (RasimKhusaenov) [inline app/graphql/types/payloads/update_project_payload.rb:5]: Why are you changing this file?

File: app/graphql/types/payloads/update_task_payload.rb (hunk 12 above)
Comment (RasimKhusaenov) [reply app/graphql/types/payloads/update_task_payload.rb:5]: @w1nway

File: app/graphql/types/user_type.rb [hunk 16]
```diff
@@ -0,0 +1,11 @@
+module Types
+  class UserType < Types::BaseObject
+    field :id, ID, null: false
+    field :first_name, String
+    field :last_name, String
+    field :email, String
+    field :created_at, GraphQL::Types::ISO8601DateTime, null: false
+    field :updated_at, GraphQL::Types::ISO8601DateTime, null: false
+    field :user_role, UserRoleEnum, null: false
```
Comment (RasimKhusaenov) [inline app/graphql/types/user_type.rb:9]: Why `user_role`? It doesn't work

**Review by RasimKhusaenov — CHANGES_REQUESTED** [review]

File: app/graphql/types/user_type.rb (hunk 16 above)
Comment (w1nway) [reply app/graphql/types/user_type.rb:9]: i think that class UserRoleEnum work only with this variable

**Pull request closed without merging**